### PR TITLE
Abstract strat contract

### PR DESCRIPTION
**Pull request by patrickmao93**

### CODE REVIEW

File: contracts/strategies/AbstractStrategy.sol [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+        if (amountFromSell < _buyAmount) {
+            uint256 buyAmount = _buyAmount - amountFromSell;
+            uint256 actualSharesFromBuy = _doBuy(buyAmount);
+            shares += actualSharesFromBuy;
+            uint256 totalSharesFromBuy = actualSharesFromBuy + _sellShares;
+            require(totalSharesFromBuy >= _minSharesFromBuy, "failed min shares from buy");
+            emit Buy(_buyAmount, totalSharesFromBuy);
+            emit Sell(_sellShares, amountFromSell);
+            return (totalSharesFromBuy, amountFromSell);
```
Comment (xiaozhou) [inline contracts/strategies/AbstractStrategy.sol:97]: redundant code, record values e.g., (`sharesFromBuy = actualSharesFromBuy + _sellShares`), then emit and return only once at the end of the function.

File: contracts/strategies/AbstractStrategy.sol [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        } else if (amountFromSell > _buyAmount) {
+            uint256 sellShares = _sellShares - sharesFromBuy;
+            uint256 actualAmountFromSell = _doSell(sellShares);
+            shares -= actualAmountFromSell / this.syncPrice();
+            uint256 totalAmountFromSell = actualAmountFromSell + _buyAmount;
+            require(totalAmountFromSell >= _minAmountFromSell, "failed min amount from sell");
```
Comment (xiaozhou) [inline contracts/strategies/AbstractStrategy.sol:103]: redundant code, record values e.g., (amountFromSell = actualAmountFromSell + _buyAmount), then emit and return only once at the end of the function.

File: contracts/strategies/AbstractStrategy.sol (hunk 1 above)
Comment (patrickmao93) [reply contracts/strategies/AbstractStrategy.sol:96]: wouldn't that change the idiom of sharesFromBuy? or I could elevate the scope of totalSharesFromBuy and totalAmountFromSell.

File: contracts/strategies/AbstractStrategy.sol [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+abstract contract AbstractStrategy is IStrategy, Ownable {
+    using SafeERC20 for IERC20;
+    using Address for address;
+
+    uint256 constant MAX_INT = 2**256 - 1;
+    uint256 constant PRICE_DECIMALS = 1e18;
+
+    address public controller;
+    address public supplyToken;
+    address public lpToken;
```
Comment (Dominator008) [inline contracts/strategies/AbstractStrategy.sol:23]: No longer needed.

File: contracts/strategies/AbstractStrategy.sol [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+            uint256 buyAmount = _buyAmount - amountFromSell;
+            uint256 actualSharesFromBuy = _doBuy(buyAmount);
+            totalSharesFromBuy = actualSharesFromBuy + _sellShares;
+            require(totalSharesFromBuy >= _minSharesFromBuy, "failed min shares from buy");
+        } else if (amountFromSell > _buyAmount) {
+            uint256 sellShares = _sellShares - sharesFromBuy;
+            uint256 actualAmountFromSell = _doSell(sellShares);
+            totalAmountFromSell = actualAmountFromSell + _buyAmount;
+            require(totalAmountFromSell >= _minAmountFromSell, "failed min amount from sell");
+            return (sharesFromBuy, totalAmountFromSell);
```
Comment (xiaozhou) [inline contracts/strategies/AbstractStrategy.sol:102]: remove this return

**Review by xiaozhou — APPROVED** [review]

**Pull request merged**